DropboxSync.UIL/Managers/ExpenseManager.cs

```diff
@@ -9,31 +17,247 @@ namespace DropboxSync.UIL.Managers
+            _mapper = mapper ??
+                throw new ArgumentNullException(nameof(mapper));
+            _expenseService = expenseService ??
+                throw new ArgumentNullException(nameof(expenseService));
+            _fileService = fileService ??
+                throw new ArgumentNullException(nameof(fileService));
+            _dropboxService = dropboxService ??
+                throw new ArgumentNullException(nameof(dropboxService));
+            _uploadService = uploadService ??
+                throw new ArgumentNullException(nameof(uploadService));
         }
 
-        public bool Create(ExpenseModelBase model)
+        // TODO : 1. Create a local backup
+        // TODO : 2. Create a dropbox backup
+        // TODO : 3. Save information in database
+        public bool Create<T>(T model) where T : ExpenseReceivedModel
         {
-            if (model.GetType() != typeof(ExpenseReceivedModel))
-                throw new ArgumentException($"To create an expense, {nameof(model)}'s type must be of " +
-                    $"type {typeof(ExpenseReceivedModel)}");
+            if (model is null) throw new ArgumentNullException(nameof(model));
+
+            // TODO : Create mapper profile from ExpenseReceivedModel to ExpenseEntity
+            ExpenseEntity expenseEntity = _mapper.Map<ExpenseEntity>(model);
+
+            if (expenseEntity is null)
+            {
+                _logger.LogError("{date} | The mapping from {modelType} to {entityType} couldn't be made!",
+                    DateTime.Now, typeof(ExpenseReceivedModel), typeof(ExpenseEntity));
+                return false;
+            }
+
+            foreach (string uploadId in model.UploadIds)
+            {
+                SavedFile? savedFile = Task.Run(async () => await _fileService.DownloadFile(uploadId)).Result;
+                if (savedFile is null)
+                {
+                    _logger.LogError("{date} | File with ID \"{fileId}\" couldn't be downloaded!", DateTime.Now, uploadId);
+                    return false;
+                }
+
+                DropboxSavedFile? dropboxSavedFile = Task.Run(async () => await _dropboxService.SaveUnprocessedFile(savedFile.FileName,
+                   DateTimeHelper.FromUnixTimestamp(model.Timestamp), savedFile.RelativePath, FileTypes.Expenses, savedFile.FileExtension)).Result;
+
+                if (dropboxSavedFile is null)
+                {
+                    _logger.LogError("{date} | File with ID : \"{fileID}\" couldn't be saved. Please read precedent logs to understand " +
+                        "the why.", DateTime.Now, uploadId);
+                    continue;
+                }
+
+                if (expenseEntity.Uploads is null) expenseEntity.Uploads = new List<UploadEntity>();
+
+                UploadEntity upload = new UploadEntity()
+                {
+                    ContentType = savedFile.ContentType,
+                    FileExtention = savedFile.FileExtension ?? "UNKNOWN",
```

Comment:
if you have the content type you definitely have the extension

Reply:
As stated previously, I put in there "in case of". If really it isn't required or any other reason I can remove it.

Reply:
file extension can be indeed null. but content type too then?

Reply:
I completly removed FileExtension property. If it is indeed needed, it will be retrieved through another way